### ISSUE TITLE
shell_exec of timelimit command hangs forever.

### ISSUE THREAD

**Issue by joelastpass** [facebook/hhvm#3506]
shell_exec, exec, proc_open etc all hang when running through hhvm, nginx->fastcgi->hhvm latest tip.

The smallest reproducible case for me is below, remove tlmelimit -q and it'll work -- php5-fpm handles this fine.  Running strace -f shows the subcommand does appear to be running properly.  

```
<?php
// Should just print 'worked',  in hhvm it hangs, leaving the processes running.
$output = shell_exec("timelimit -q /bin/echo worked");
echo $output;
```

**Comment by joelastpass** [comment]
FYI still an issue in 3.5.0 running Ubuntu 14.04, apparmor jailed, nginx -> fastcgi can't think of anything else relevant: 

HipHop VM 3.5.0 (rel)
Compiler: tags/HHVM-3.5.0-0-ga90f4733cfa0d8fefdafc7433f758f78cdc75424
Repo schema: b88b6fee8ef78ab638506d212abcbefe74a276ac

**Comment by jwatzman** [comment]
I can reproduce on a very stock 14.04 install with the official 3.5 package (i.e., the one you listed above). Not sure when anyone can get a chance to look at it.

**Comment by jwatzman** [comment]
Gah this was annoying, but I know what's going on.

- HHVM is hanging waiting on a `read` syscall from the `timeout` child, which is still running.
- The `timeout` child hasn't noticed that `/bin/echo` finished because `SIGCHLD` is blocked. The right way to wait for a child to exit is to use `sigsuspend` to make sure that the signals you're waiting on are unblocked. `timeout` doesn't do this. It calls `pause`, without checking for `SIGCHLD` being unblocked or anything like that. (And that would be wrong too, if you use `sigprocmask` this way you probably have a race condition.) Dear author of `timeout`: remember when way back in university your professor told you never ever to use `pause` and that it's always bad? This is why.j
- This doesn't show up in other environments -- command line or FB's built-in webserver -- because AFAICT it's `nginx` or something in there blocking `SIGCHLD`. It probably shouldn't be doing that, either -- but well-behaved programs should deal with that just fine, see above.
- HHVM already makes sure to restore the default signal handler for `SIGCHLD`; we can probably do the same with the signal mask to work around this. Sigh.

Seeing `SIGCHLD` blocked is easy to see if you run this C program under `shell_exec`:

```c
#define _GNU_SOURCE
#include <string.h>
#include <signal.h>
#include <stdio.h>

int main(void) {
	sigset_t mask;
	sigprocmask(SIG_BLOCK, NULL, &mask);
	if (sigismember(&mask, SIGCHLD)) {
		printf("Child is blocked!\n");
	} else {
		printf("Child is unblocked!\n");
	}
	return 0;
}
```

**Comment by jwatzman** [comment]
Not spending any more time on this now. The signal blocking is [coming from proxygen](https://github.com/facebook/proxygen/blob/dffee0440bced41a0f74e92dc1572f2f4d61ffd7/proxygen/lib/services/WorkerThread.cpp#L96-L110) when starting up an HHVM worker thread, and it's unclear why it does that. HHVM *probably* doesn't need it, but need to plumb through proxygen disabling that behavior -- or figure out a nice way to do it -- and not worth the time right now IMO.

You can work around this by using [`pcntl_sigprocmask`](http://php.net/manual/en/function.pcntl-sigprocmask.php) to unblock `SIGCHLD` -- or by using a program that isn't broken like `timelimit` is.

**Comment by joelastpass** [comment]
Nice detective work.  I can confirm that the workaround works so agree that it's low-priority.

For future searchers, adding this will allow timelimit to work:

pcntl_sigprocmask(SIG_UNBLOCK, array(SIGCHLD));

**Comment by cpollett** [comment]
I ran into I think a similar issue. I installed hhvm today using apt-get and am running apache and fast-cgi. Consider the following three files:
test2.php:
<?php
file_put_contents('/var/www/yo.txt', 'hi there');
?>
test1a.php
<?php
echo "Fork Test";
$fork = "/usr/bin/hhvm -f '/var/www/test2.php' < /dev/null > /dev/null &";
pclose(popen($fork, "r"));
?>
test1b.php
<?php
echo "Fork Test";
$fork = "/usr/bin/php -f '/var/www/test2.php' < /dev/null > /dev/null &";
pclose(popen($fork, "r"));
?>
If I run test1a.php or test1b.php from the command line it makes the file yo.txt as expected.
The web server has permissions on /var/www. When I run test1a.php from the server
it doesn't work, but test1b.php does. The
pcntl_sigprocmask(SIG_UNBLOCK, array(SIGCHLD));
suggestion did not make it work.

Best,
Chris

**Comment by jwatzman** [comment]
@cpollett if `pcntl_sigprocmask(SIG_UNBLOCK, array(SIGCHLD));` doesn't fix it then it's a different problem. Please file a new issue and we can help you there.
